### DISCUSSION TITLE
Best Approach for Rapid Storage Access

### DISCUSSION THREAD

**Discussion by Butterfield-Isabelle** [supabase/supabase#7183] [Questions] [answered]
Hi! I'm looking for some guidance on how I can get the fastest consistent access to my items in storage. I'm aware of the key access methods - from.download() and via public URL - but I'm encountering slight obstacles with each, which I'm hoping to overcome. Here is some context:

I'm making a little YouTube-like application - the relevant similarities are in the user accounts and the comment functionality. When a user is signed in, they should always have their profile image showing in the nav bar, and whenever viewing comments, each comment should display the profile image of the user that commented. These images obviously need to be pulled from storage, and here's what I'm encountering with the two methods:

**from.download()** - I've been trying to keep these database calls to a minimum, particularly because they seem to be taking a while. But maybe I'm doing something wrong. For example, I have an "update user info" function that contacts the database 3 times and takes about 4 seconds to complete, which feels a little too long. The sequence consists of an image upload to storage, followed by a request for the signed url of that image, and concludes with an update to the table with user info. Is it surprising that these calls take so long? Maybe it's just because the upload takes longer than other database calls? It's making me shy away from using many from.download() calls, especially when I need to render a bunch of profile images on an array of comments quickly.

**public URL** - This access method is very convenient and seems faster, but it is causing problems when the URL intermittently changes. I believe the "token" component that tacks on to the end of the URL changes every day or so - does this sound right? The way I was intending to use the URL is to store the string in each user's info table so that when I pull their profile info, I can have immediate access to their profile image via the public URL. Is there a way I can keep the public URL in the user's table up-to-date with the actual public URL of the item in storage?

Perhaps one or both of these current approaches are in violation of best practices - I would love any guidance available. Please let me know if this could use further clarification. Thank you!

**Comment by GaryAustin1** [comment, marked as answer]
If you call getPublicUrl it does not make a database call, so is very quick.  It just puts together the path name based on constants and path info.
There is no token added by the call.   The dashboard storage does add a "cachebuster" token when you use it to get the public URL, but it is not used in getting the actual file path so when it changes it still gets the same file.

**Reply by Butterfield-Isabelle** [reply]
Thanks for the reply! Interesting - maybe I've misdiagnosed my issue then. The result I'm seeing, which made me conclude that the above issue was happening, is that after ~24 hours, when I try to render the same image URL that has been working for the last little while, I suddenly get a JWT error. Does that ring a bell for another issue, possibly?

**Reply by GaryAustin1** [reply]
I don't remember the error if the createSignedUrl() url, with a token that expires per your parameter passed to it, expires.
But public URL has no expiration or JWT.  It goes straight to the AWS storage.
You might need to provide more info on the error, how you generated the URL and what is going on when you call it.

**Comment by thorwebdev** [comment]
For avatars we recommend creating a public bucket and storing the public URL in the profiles table.

**Reply by Butterfield-Isabelle** [reply]
Maybe this is related - I would like to create a public bucket, but Supabase isn't accepting the syntax from the policy template for a public bucket, so right now it is restricted to authenticated users only. Here is what happens - in the template, I leave the Target Roles blank, the template inserts "public" into the SQL query, but the system doesn't like something about the syntax and I don't have control over editing the SQL to try and resolve, since it's a read-only editor.

<img width="1151" alt="Screen Shot 2022-06-07 at 4 45 38 AM" src="https://user-images.githubusercontent.com/96531515/172372224-85e79c90-69a3-4924-80ac-5540b9c059a7.png">
<img width="1440" alt="Screen Shot 2022-06-07 at 4 45 54 AM" src="https://user-images.githubusercontent.com/96531515/172372303-e2b2fd00-8667-4457-9ee3-5783082d2468.png">

**Comment by GaryAustin1** [comment]
First a public bucket is set by selecting from the option menu icon you see when you select the bucket.
![image](https://user-images.githubusercontent.com/54564956/172396566-624644da-9e80-47cb-bdd4-dc63a0161ab2.png)

I'm able set policies on a public bucket.

Normally when you have policies on the object table you need to also filter on the bucket name:
```
(bucket_id = 'test'::text) AND ......
```

**Reply by Butterfield-Isabelle** [reply]
That was the issue - I didn't realize there was a difference between a bucket being public and its policies allowing public access. So I had been using the createSignedURL() function which was timing out, instead of the getPublicURL() function which wouldn't work until the bucket was public. Thank you!

**Reply by Prudence97** [reply]
Thank you for asking a similar question. I've been thinking about this as I've made plans to work on an upload file codes.

I think think this answers in some way
